### DISCUSSION TITLE
Authentication Session

### DISCUSSION THREAD

**Discussion by justmangoou** [maxcountryman/tower-sessions#11] [Ideas]
I believe that most people use this for handling authentication session (aka login session) which it's not really implemented here, should this become a feature. If yes,

- Add this feature as lib feature
- Allow user to create session with provided id
- Allow session to be generated manually

**Comment by justmangoou** [comment]
@maxcountryman what do you think of this?

**Reply by maxcountryman** [reply]
Thanks for mentioning this.

I'm wondering if `axum-login` should migrate to use this crate. (I started [a discussion](https://github.com/maxcountryman/axum-login/discussions/100) over there about this too.)

Alternatively we could implement a `tower-login` but I'm not sure if that should live separately from axum.

I suspect it's probably best to have this life as its own crate that consumes this crate tho.

**Comment by justmangoou** [comment]
Well, it's actually easier to build something from scratch than fix the existing one. But it's up to you to decide.

However, not every user chooses to use a password-based login for their application. They can use oauth or openid which is easier for them to interact with 3rd services such as Github. Also, many of them tend to handle sessions manually.

So that made me think adding such a "general authentication session" feature could handle both cases perfectly.

P/s: I didn't see the reply box

**Reply by maxcountryman** [reply]
> However, not every user chooses to use a password-based login for their application. They can use oauth or openid which is easier for them to interact with 3rd services such as Github.

This is a good point.

Do you have a design in mind?

**Reply by justmangoou** [reply]
For ``tower-sessions``, there's no need to change anything except for those points I've mentioned.

For ``axum-sessions`` or ``tower-sessions`` (if plan to), just need to replace the password-input flow with oauth's or openid's flow.

I looked into Ory, Casbin. It provides many features that I personally don't need and it costs performance. So I think just keep it as simple as possible.

**Comment by justmangoou** [comment]
For further development,
https://github.com/maxcountryman/axum-login/discussions/52 and some other discussions about implementing non-supported database drivers or ORMs/ODMs
https://github.com/maxcountryman/axum-login/discussions/59 - this explained why authentication session needs to be given manually or any method that allows to renew session dynamically.

In spite of using ``tower`` as the based of this library, some ``axum``'s feature may be considerable to be put in this library such as ``CookieJar``. I'm not certain that this makes any different compares to ``tower``, so I will leave this question.


~~For the design, something like~~
```rust
let session_manager = AuthSessionManagerLayer::new(session_store).with_secure(true);
```

```rust
#[derive(Debug, Clone)]
pub struct AuthSessionManagerLayer<Store: SessionStore> {
    inner: SessionManagerLayer<Store>,
}
```
This should be applied to login crate

Edited design:
- Allow construct ``SessionManager`` publicly.
- Add SessionManager to State (idk if tower has this) and retrieve it from middleware.
- Add option to allow ``SessionManager`` issues session automatically or not.

I might miss some features, but in 99% way to handle this, it will cause breaking change.

**Comment by justmangoou** [comment]
This should be implemented by authentication libraries instead.

**Reply by maxcountryman** [reply]
That makes sense to me.

**Reply by justmangoou** [reply]
However, I think SessionManager is still needed to be extracted.

Is it possible to extract SessionManager from http method like Session? It'd better than the latest PR, no breaking changes, no need to pass SessionManager parameter.

**Reply by maxcountryman** [reply]
I've put together a functional proof-of-concept [over here](https://github.com/maxcountryman/axum-login/discussions/100#discussioncomment-7188499).
